destination/format/csv.go

```diff
@@ -0,0 +1,188 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package format
+
+import (
+	"bytes"
+	"encoding/csv"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-errors/errors"
+	"golang.org/x/exp/maps"
+)
+
+func makeCSVRecords(records []sdk.Record, prefix string, orderingColumns []string) (
+	*bytes.Buffer, *bytes.Buffer, map[string]string, []string, []string, error,
+) {
+	var (
+		insertsBuf bytes.Buffer
+		updatesBuf bytes.Buffer
+	)
+	insertsWriter := csv.NewWriter(&insertsBuf)
+	updatesWriter := csv.NewWriter(&updatesBuf)
+
+	// we need to store the operation in a column, to detect updates & deletes
+	operationColumn := fmt.Sprintf("%s_operation", prefix)
+	columnMap := map[string]string{operationColumn: "VARCHAR"}
+	csvColumnOrder := []string{operationColumn}
+	// TODO: see whether we need to support a compound key here
+	// TODO: what if the key field changes? e.g. from `id` to `name`? we need to think about this
+
+	for _, r := range records {
+		// get Primary Key(s)
+		if len(orderingColumns) == 0 {
+			var recordKeyMap map[string]interface{}
+			// we are making an assumption here that it's structured data
+			if err := json.Unmarshal(r.Key.Bytes(), &recordKeyMap); err != nil {
+				return nil, nil, nil, nil, nil,
+					errors.Errorf("could not unmarshal record.key, only structured data is supported: %w", err)
+			}
+			orderingColumns = maps.Keys(recordKeyMap)
+		}
+
+		var a sdk.Data
+		// create a column map if we are updating or creating records
+		if r.Operation != sdk.OperationDelete {
+			a = r.Payload.After
+		} else {
+			a = r.Payload.Before
+		}
+
+		// infer ordering column from first record
+		var cols map[string]interface{}
+
+		// we are making an assumption here that it's structured data
+		if err := json.Unmarshal(a.Bytes(), &cols); err != nil {
+			return nil, nil, nil, nil, nil,
+				errors.Errorf("could not unmarshal record.payload.after, only structured data is supported: %w", err)
+		}
+
+		for key, val := range cols {
+			if columnMap[key] == "" {
+				csvColumnOrder = append(csvColumnOrder, key)
+				switch val.(type) {
+				case int, int8, int16, int32, int64:
+					columnMap[key] = "INTEGER"
+				case float32, float64:
+					columnMap[key] = "INTEGER"
```

Comment:
That's weird, integers also accept float numbers in snowflake?

Reply:
good catch - I don't think so. we added this because we countered a scenario that was the other way around, an integer being placed into a float.

that said, interpreting numbers as int vs float is a tricky problem, so I think we should just broadly use `NUMBER`. @anna-cross @lyuboxa feel free to chime in

https://docs.snowflake.com/en/sql-reference/data-types-numeric

Reply:
This is fine, I'm going to fix this to emit the proper types.